### ISSUE TITLE
Most wanted / upcoming features and bugfixes

### ISSUE THREAD

**Issue by pfink** [openhab/openhab-website#12]
In the community, we repeatedly discussed the concept of a "roadmap" (e.g. [here](https://community.openhab.org/t/future-of-openhab/35341)). Many people would like to see an overview of "most wanted" featured and also about upcoming features for the next version. I want to clearly state that, of course, nobody can enforce or guarantee any progress on certain issues within a specified time (so please, let's exclude this discussion here).

But it clearly would help to bring focus on topics that many people are interested in. An issue that already interests many people, will likely also be a point of interest for others who didn't saw the issue yet. The average user is not expert of the openHAB architecture, so he don't know which repositories exist, where too look, how to filter, and so on. For those people (and I'm sure also for technical users) it would be cool to have two dynamic inline overviews on the website:

-------
**Most wanted features**

Query for all open issues and PRs, descending by +1 reactions / upvotes: 

- openHAB (full organization search): https://github.com/issues?utf8=✓&q=is%3Aopen+user%3Aopenhab+sort%3Areactions-%2B1-desc
- ESH: https://github.com/eclipse/smarthome/issues?utf8=✓&q=is%3Aopen+sort%3Areactions-%2B1-desc+

Both queries could be aggregated / merged using the [search api](https://developer.github.com/v3/search/#search-repositories).

**Upcoming features**

Here we would have merged PRs since the last release, preferably also ordered descending by +1 reactions. For filtering the right PRs, probably the same criteria can be used like for @BClark09's changelog:  https://gist.github.com/BClark09/6f13345f0a7f38482b444b47c397a1b4

-------

Voting: Currently, the built-in GitHub vote mechanisms are not used so often, and they're also not used consistently. It would be good to give hands-on advice on the website how to vote for an issue or (even better) use some GitHub integration / API to make it possible directly voting for an issue / PR on the website.

**Comment by kaikreuzer** [comment]
Voting sounds rather awful to me as it implicitly suggests to people that this will influence any priority list of "someone" who implements the stuff for them.
Instead of promoting such a passive attitude, we should imho rather show a list of issues labels as [good first issue](https://help.github.com/articles/helping-new-contributors-find-your-project-with-labels/) to help people find a way to join the active development of the project, possibly on the page https://website.openhab.org/about/contributing.html#contribute-code.

Wrt the release notes / upcoming features: I think this is something we should look at in the context of doing regular milestones - yes, they should be listed somewhere, including their generated release notes. With that, even before a release, people will see what is (definitely) going to be in the next release.

Wrt a roadmap, at ESH we trying to keep a [rough backlog](https://github.com/eclipse/smarthome/projects/3) of things that are considered to be important to work on next. But not all of that might be relevant to openHAB or to end users, so I wouldn't plainly show this anywhere on the openHAB website.

**Comment by pfink** [comment]
> Voting sounds rather awful to me as it implicitly suggests to people that this will influence any priority list of "someone" who implements the stuff for them.

I don't think that such thing is suggested. Anyhow, the priority is actually increasing. How? Because the higher an issue is voted, the better is the visibility of this issue. The better the visibility, the higher the number of people who are aware of this issue. The higher the number of people aware of a certain issue, the more likely it is that one or more of these people will work on it. An issue that was never seen by any developer at all, has a chance of 0% of making progress.

> we should imho rather show a list of issues labeled as good first issue to help people find a way to join the active development of the project, possibly on the page https://website.openhab.org/about/contributing.html#contribute-code.

I don't think that it works this way. People mostly want to work on things they're personally interested in and not something that was "randomly" picked by a maintainer. Issues that already attracted many people, will also more likely be attractive for a developer. Maybe a "good first issue" can help a new developer to make some exercise before starting with real topics he is interested in, but it won't motivate starting to develop at all.

> But not all of that might be relevant to openHAB or to end users, so I wouldn't plainly show this anywhere on the openHAB website.

An issue that is not relevant to any end user very likely won't have enough votes to show up on the website.

**Comment by ghys** [comment]
Please be advised there is a harsh rate limit on the GitHub API for unauthenticated requests (according to https://developer.github.com/v3/#rate-limiting, 60 requests/hr).
After a few calls in several seconds, it will refuse further queries.

To surface popular feature requests, Discourse has a [voting plugin](https://meta.discourse.org/t/discourse-voting/40121), perhaps it would help keeping the end user feature requests out of GitHub issues and avoid them being submitted to the wrong repo.
(FWIW, Home Assistant has a [feature requests](https://community.home-assistant.io/c/feature-requests/l/votes) category with voting activated on their forum.)

**Comment by pfink** [comment]
The website could just cache the result, it does not have to make an API request for every page view, does it?

**Comment by ghys** [comment]
It's a bunch of static pages generated at each deployment with so server-side, so unless it's cached in the browser's local storage (which has other problems), not really.

**Comment by pfink** [comment]
But couldn't we easily add a server-side, e.g. a cronjob that updates the static page hourly?

**Comment by kaikreuzer** [comment]
> (FWIW, Home Assistant has a feature requests category with voting activated on their forum.)

That sounds like a pretty good option to me, too!

**Comment by pfink** [comment]
You mean carving out the feature requests from GitHub issues and migrating them to discourse? Mhm, I'm not convinced...

**Comment by ThomDietrich** [comment]
I like the general idea here and the suggestion of a discourse category.

@pfink GitHub feature requests are imho a totally different area. All GitHub activity should ideally be reserved for developers. Users most often don't contribute meaningful issues. A Forum category would offer the opportunity for end users to explain their request in a user-perspective. Perfect fit for what you asked for.

Btw: One could always resort to uservoice.com...

**Comment by Confectrician** [comment]
+1 for a vote based discourse solution too.
I have mixed feelings about feature requests in a GitHub tool called "issues" for a longer time.

A bit Offtopic:
Nevertheless i like the idea of providing some GitHub-Contents/Stats on the website.
I had a similar idea about showing our repo maintainers based on the organisation team-members somehow.

The API has a request Limit of 5000 at the moment you authenticate on some way, which should be possible during a cronjob or ci job that starts a script for updating the wanted values.

**Comment by kaikreuzer** [comment]
> I had a similar idea about showing our repo maintainers based on the organisation team-members somehow.

That's not the topic of this issue, but I was thinking of such a page myself for a long time - to easily show, who are the maintainers of which repo and make it more personal by not only listing github handles, but also showing their profile pictures from Github on a nicely rendered page.
To avoid github rate limits, it should be ok to pre-generated this info in our groovy scripts that also assemble all the add-on infos.

**Comment by pfink** [comment]
> GitHub feature requests are imho a totally different area. All GitHub activity should ideally be reserved for developers. Users most often don't contribute meaningful issues. A Forum category would offer the opportunity for end users to explain their request in a user-perspective. Perfect fit for what you asked for.

I see it a bit different. I wanted to have a easy-accessible ranking for all concrete issues & feature requests, not for "not meaningful issues explained from a user perspective". I think for those requests it works already right now, the forum offers the possibility for users to explain requests from a user perspective. Other (maybe more expierienced) users can support and eventually point out whether a GitHub issue makes sense.

**Comment by ThomDietrich** [comment]
Interesting that we seemingly have the same goal but completely different opinions on the solution. I can only repeat my arguments:

* **GitHub:** developer oriented, implementation oriented, technical
* **Community:** end user oriented, functional/non-functional, non-technical

If the goal is to create a place for non-tech-savvy end users to discuss broad ideas and wishes (e.g. "Provide an alternative icon set" or "Offer a setup routine for platform X") they should be able to discuss it freely, discuss it openly with other end users, and should not have to bother with questions like which repo the issue belongs to or which technical questions need to be answered. 
The new category would not be a "parallel world" but rather an area of the community dedicated to non-technical proposals, ideas and wishes.

**Comment by gitMiguel** [comment]
> > (FWIW, Home Assistant has a feature requests category with voting activated on their forum.)
> 
> That sounds like a pretty good option to me, too!

@kaikreuzer Is it possible to open this kind of voting category? Especially now as OH v.3 is coming.
